### DISCUSSION TITLE
`klog edit` fails if `$EDITOR` contains flags

### DISCUSSION THREAD

**Discussion by hovsater** [jotaen/klog#355] [Bugs]
When I run `klog edit` I get the following error message:

```console
Error: Cannot open preferred editor
$EDITOR variable was: subl -w
```

There's nothing wrong with my `$EDITOR` variable and other tools such as `git` open it just fine.

**Comment by jotaen** [comment]
From what I know, `$EDITOR` is actually not supposed to contain arguments (such as `-w`), but it’s rather treated as a single word. (See [1](https://superuser.com/questions/521070/unix-environment-variables-with-arguments), [2](https://stackoverflow.com/questions/8299610/zsh-command-not-found-for-editor), though it also seems to be depending on the shell.) Therefore, klog wouldn’t expand `subl -w` in the sense that it would feed `-f` to the `subl` command, but it interprets the whole string as if was one “atomic” command.

The problem is that it’s not safe to split the string, because it can legitimately contain spaces, such as `/home/Foo Bar/bin/subl`. There was actually also [a recent issue in regards to that](https://github.com/jotaen/klog/issues/206). Maybe there is a way to fix this, though, if git manages to do it too, so I suppose that’s worth to check out.

In any event, I suppose the error message could be improved further, so if there is a space character in the `$EDITOR` variable, it could warn you that flags arguments cannot be processed.

**Comment by hovsater** [comment]
@jotaen I'm not sure that's accurate. I've never stumbled upon a Unix program that doesn't correctly handle spaces within `$EDITOR`. Do you have any "official" resources regarding this other than the the ZSH specific ones above? I could of course wrap `subl -w` using a bash script, but I much rather not.

**Comment by hovsater** [comment]
Looking into this further, there seem to be no consensus in how to handle this. Most program **does** handle this, but since there is no specification, you can't be certain. 

> The problem is that it’s not safe to split the string, because it can legitimately contain spaces, such as /home/Foo Bar/bin/subl. There was actually also https://github.com/jotaen/klog/issues/206. Maybe there is a way to fix this, though, if git manages to do it too, so I suppose that’s worth to check out.

I guess a solution would be to quote the parts that should stay together. For example:

```shell
export EDITOR="'/a/path/containing spaces/editor'"
export EDITOR="'/a/path/containing spaces/editor' -arg"
```

I'll leave it up to you whether or not you want to pursue that. 🙂 If not, I'll just have to create a wrapper script.

**Comment by jotaen** [comment]
Your suggestion with quoting seems to be the cleanest to me, because it’s bash-native and leaves it to the user to choose from both options. I guess the error message would still need to be improved then, in case people aren’t aware of the quoting option.

The downside is that it requires to parse the `$EDITOR` string properly, because the quotes could be escaped and such. Potential libraries would be:

- [`shellquote`](https://github.com/kballard/go-shellquote): “old” but seems to do its job
- [`shellwords`](https://pkg.go.dev/github.com/mattn/go-shellwords): [doesn’t properly support escaping on Windows](https://github.com/mattn/go-shellwords/issues/38), which is likely a dealbreaker. Also has more fancy features, such as variable expansion, which we don’t need.

tldr, I’m open to fixing this, but it’s not totally trivial.
